ext/standard/array.c

```diff
@@ -1233,15 +1233,56 @@ PHP_FUNCTION(min)
 		}
 	} else {
 		/* mixed min ( mixed $value1 , mixed $value2 [, mixed $value3... ] ) */
-		zval *min, result;
-		int i;
+		zval *min;
+		uint32_t i;
 
 		min = &args[0];
+		zend_long min_lval;
+		double min_dval;
 
-		for (i = 1; i < argc; i++) {
-			is_smaller_function(&result, &args[i], min);
-			if (Z_TYPE(result) == IS_TRUE) {
-				min = &args[i];
+		if (Z_TYPE_P(min) == IS_LONG) {
+			min_lval = Z_LVAL_P(min);
+
+			for (i = 1; i < argc; i++) {
+				if (EXPECTED(Z_TYPE(args[i]) == IS_LONG)) {
+					if (min_lval > Z_LVAL(args[i])) {
+						min_lval = Z_LVAL(args[i]);
+						min = &args[i];
+					}
+				} else if (Z_TYPE(args[i]) == IS_DOUBLE) {
+					min_dval = (double) min_lval;
+					goto double_compare;
```

Comment:
Is this really safe? As currently written, when a double is observed, all comparions are done using double, which might compare different longs as the same value: https://3v4l.org/EgniH/rfc#vgit.master_jit

The original issue is primary because of tracing overhead, IMHO microoptimization for <10% is not much helpful vs. the much worse code.

Reply:
@mvorisek Note that the zend_compare function (which was previously always used) also casts to a double. I created a test case for min: https://3v4l.org/ZsosL and this behaves the same with this patch and without this patch. It is indeed true that there is a loss of precision when comparing very big doubles, but this is already the case now. There is afaik no change in behaviour.

Or did you test this patch and found an issue? If so, please provide the reproducer.

> The original issue is primary because of tracing overhead, IMHO microoptimization for <10% is not much helpful vs. the much worse code.

It turns out that this actually improves the diff real-world workload from the original issue report a lot. The benchmarks I did were only very isolated but it's important to look at the real-world workload. Also there are already a couple of functions which have a fast-path for longs & doubles iirc.

Reply:
https://github.com/php/php-src/blob/24771fb08b/ext/standard/array.c#L1267 will produce a bad result when there are doubles between large integers, I did not tested it, but originally there was no jump to double cmp only

Reply:
> https://github.com/php/php-src/blob/24771fb08b/ext/standard/array.c#L1267 will produce a bad result when there are doubles between large integers, I did not tested it, but originally there was no jump to double cmp only

I see what you mean. Here's a test with a behaviour difference:

```php
<?php

var_dump(PHP_INT_MAX*3);
var_dump(min(PHP_INT_MAX*2, PHP_INT_MAX, PHP_INT_MAX-1));
// int(9223372036854775806) previously
// int(9223372036854775807) with this patch
```

This can probably be worked around. I would prefer a solution that doesn't restart the loop to prevent a performance impact. I also checked with my generic zend_compare improvement and it doesn't suffer from that problem, so we still have that as a possibility.

> the issue (in the diff lib) will be still present, at the main slowdown is due tracing overhead, not due slow php-src

Yes, but I didn't say it fully got rid of all the overhead, I said this is a big improvement.